datasources-agroal/src/main/java/org/wildfly/extension/datasources/agroal/AbstractDataSourceDefinition.java

```diff
@@ -0,0 +1,247 @@
+            .build();
+
+    static final SimpleAttributeDefinition STATISTICS_ENABLED_ATTRIBUTE = create("statistics-enabled", ModelType.BOOLEAN)
+            .setAllowExpression(true)
+            .setDefaultValue(new ModelNode(false))
+            .setRequired(false)
+            .build();
+
+    // --- connection-factory attributes //
+
+    static final SimpleAttributeDefinition DRIVER_ATTRIBUTE = create("driver", ModelType.STRING)
+            .setRequires()
+            .setRestartAllServices()
+            .setValidator(new StringLengthValidator(1))
+            .build();
+
+    static final SimpleAttributeDefinition URL_ATTRIBUTE = create("url", ModelType.STRING)
+            .setAllowExpression(true)
+            .setRestartAllServices()
+            .setRequired(false)
+            .setValidator(new StringLengthValidator(1))
+            .build();
+
+    static final SimpleAttributeDefinition TRANSACTION_ISOLATION_ATTRIBUTE = create("transaction-isolation", ModelType.STRING)
+            .setAllowExpression(true)
+            .setAllowedValues(NONE.name(), READ_UNCOMMITTED.name(), READ_COMMITTED.name(), REPEATABLE_READ.name(), SERIALIZABLE.name())
+            .setRequired(false)
+            .setRestartAllServices()
+            .setValidator(EnumValidator.create(AgroalConnectionFactoryConfiguration.TransactionIsolation.class, EnumSet.allOf(AgroalConnectionFactoryConfiguration.TransactionIsolation.class)))
+            .build();
+
+    static final SimpleAttributeDefinition NEW_CONNECTION_SQL_ATTRIBUTE = create("new-connection-sql", ModelType.STRING)
+            .setAllowExpression(true)
+            .setRequired(false)
+            .setRestartAllServices()
+            .build();
+
+    static final SimpleAttributeDefinition USERNAME_ATTRIBUTE = create("username", ModelType.STRING)
+            .setAttributeGroup("security")
+            .setAllowExpression(true)
+            .setRequired(false)
+            .setRestartAllServices()
+            .setValidator(new StringLengthValidator(1))
+            .build();
+
+    static final SimpleAttributeDefinition PASSWORD_ATTRIBUTE = create("password", ModelType.STRING)
+            .setAttributeGroup("security")
+            .setAllowExpression(true)
+            .setRequired(false)
+            .setRestartAllServices()
+            .setValidator(new StringLengthValidator(1))
+            .build();
+
+    static final PropertiesAttributeDefinition CONNECTION_PROPERTIES_ATTRIBUTE = new PropertiesAttributeDefinition.Builder("connection-properties", true)
+            .setAllowExpression(true)
+            .setRequired(false)
+            .setRestartAllServices()
+            .build();
+
+    static final ObjectTypeAttributeDefinition CONNECTION_FACTORY_ATTRIBUTE = ObjectTypeAttributeDefinition.create("connection-factory", DRIVER_ATTRIBUTE, URL_ATTRIBUTE, TRANSACTION_ISOLATION_ATTRIBUTE, NEW_CONNECTION_SQL_ATTRIBUTE, USERNAME_ATTRIBUTE, PASSWORD_ATTRIBUTE, CONNECTION_PROPERTIES_ATTRIBUTE)
```

Comment:
Are we sure we want to have "url" as common attribute for connection factory for all, driver, datasource and xa-datasource? I see in agroal that if driver class is javax.sql.Datasource or javax.sql.XADatasource, "url" is injected using "setUrl" or "setProperty". I'm not sure if it's good idea. There are drivers that haven't such setters and use "serverName", "port" and "databaseName" instead of "url". In such case agroal throws NoSuchMethodException which ends as WARN in WF log.

Reply:
ok, I'll look into this.